In [7890]:
import numpy as np
import pandas as pd
import seaborn as sns

In [7891]:
df=pd.read_csv('../Data/Credit_score.csv',low_memory=False)


In [7892]:
pd.options.display.max_columns=30

In [7893]:
df.head(8)

ID Customer_ID     Month           Name   Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2  0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3  0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4  0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   
5  0x1607   CUS_0xd40      June  Aaron Maashoh    23  821-00-0265  Scientist   
6  0x1608   CUS_0xd40      July  Aaron Maashoh    23  821-00-0265  Scientist   
7  0x1609   CUS_0xd40    August            NaN    23    #F%$D@*&8  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0      19114.12            1824.843333                  3                4   
1      19114.12                    NaN                  3                4   
2      19114.12                    NaN                  3                4   
3      19114.12                    NaN                  3                4   
4      19114.12            1824.843333                  3                4   
5      19114.12                    NaN                  3                4   
6      19114.12            1824.843333                  3                4   
7      19114.12            1824.843333                  3                4   

   Interest_Rate Num_of_Loan  \
0              3           4   
1              3           4   
2              3           4   
3              3           4   
4              3           4   
5              3           4   
6              3           4   
7              3           4   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                   -1   
2  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
3  Auto Loan, Credit-Builder Loan, Personal Loan,...                    5   
4  Auto Loan, Credit-Builder Loan, Personal Loan,...                    6   
5  Auto Loan, Credit-Builder Loan, Personal Loan,...                    8   
6  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
7  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   

  Num_of_Delayed_Payment Changed_Credit_Limit  Num_Credit_Inquiries  \
0                      7                11.27                   4.0   
1                    NaN                11.27                   4.0   
2                      7                    _                   4.0   
3                      4                 6.27                   4.0   
4                    NaN                11.27                   4.0   
5                      4                 9.27                   4.0   
6                     8_                11.27                   4.0   
7                      6                11.27                   4.0   

  Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  \
0          _           809.98                 26.822620   
1       Good           809.98                 31.944960   
2       Good           809.98                 28.609352   
3       Good           809.98                 31.377862   
4       Good           809.98                 24.797347   
5       Good           809.98                 27.262259   
6       Good           809.98                 22.537593   
7       Good           809.98                 23.933795   

      Credit_History_Age Payment_of_Min_Amount  Total_EMI_per_month  \
0  22 Years and 1 Months                    No            49.574949   
1                    NaN                    No            49.574949   
2  22 Years and 3 Months                    No            49.574949   
3  22 Years and 4 Months                    No            49.574949   
4  22 Years and 5 Months                    No            49.574949   
5 

In [7894]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      90015 non-null   object 
 4   Age                       100000 non-null  object 
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  object 
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

In [7895]:
df.describe()

Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
count           84998.000000      100000.000000     100000.00000   
mean             4194.170850          17.091280         22.47443   
std              3183.686167         117.404834        129.05741   
min               303.645417          -1.000000          0.00000   
25%              1625.568229           3.000000          4.00000   
50%              3093.745000           6.000000          5.00000   
75%              5957.448333           7.000000          7.00000   
max             15204.633330        1798.000000       1499.00000   

       Interest_Rate  Delay_from_due_date  Num_Credit_Inquiries  \
count  100000.000000        100000.000000          98035.000000   
mean       72.466040            21.068780             27.754251   
std       466.422621            14.860104            193.177339   
min         1.000000            -5.000000              0.000000   
25%         8.000000            10.000000              3.000000   
50%        13.000000            18.000000              6.000000   
75%        20.000000            28.000000              9.000000   
max      5797.000000            67.000000           2597.000000   

       Credit_Utilization_Ratio  Total_EMI_per_month  
count             100000.000000        100000.000000  
mean                  32.285173          1403.118217  
std                    5.116875          8306.041270  
min                   20.000000             0.000000  
25%                   28.052567            30.306660  
50%                   32.305784            69.249473  
75%                   36.496663           161.224249  
max                   50.000000         82331.000000

# ID, Customer_ID, Month, Name, SSN

#### -> There are 12500 Customers. Each Customer has data for 8 Months and hence we have data for 100000 rows.

#### -> 5572 SSN values are not in the correct format.

#### -> Name Column has 9985 missing Values and 10139 Unique Name. Few Customers may have similar Names.

#### -> Imputing the missing Name values as Unknown.


In [7896]:
df.Customer_ID.nunique()

12500

In [7897]:
df.ID.nunique()

100000

In [7898]:
df.Month.unique()

array(['January', 'February', 'March', 'April', 'May', 'June', 'July',
       'August'], dtype=object)

In [7899]:
df.SSN.nunique()

12501

In [7900]:
df.SSN.value_counts()

#F%$D@*&8      5572
078-73-5990       8
486-78-3816       8
750-67-7525       8
903-50-0305       8
               ... 
856-06-6147       4
753-72-2651       4
331-28-1921       4
604-62-6133       4
286-44-9634       4
Name: SSN, Length: 12501, dtype: int64

In [7901]:
df.Name.isna().sum()

9985

In [7902]:
df.Name.nunique()

10139

In [7903]:
df.Name=df.Name.str.lower()

In [7904]:
sample=df.merge(df.groupby('Customer_ID')['Name'].agg(pd.Series.mode).reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7905]:
df.Name=np.where(df.Name.isna(),sample.Name_y,df.Name)

In [7906]:
df.Name.nunique()

10139

In [7907]:
df.Name.isna().sum()

0

In [7908]:
df.Name.value_counts()

langep                48
jessicad              48
stevex                48
huwk                  40
jonathan stempelr     40
                      ..
harro teni             8
breidthardtj           8
sven egenterx          8
antonella ciancioc     8
nicks                  8
Name: Name, Length: 10139, dtype: int64

# Age 

#### -> Age Column has some large values. 

#### -> Each Customer has 8 rows of data for 8 months. We are assuming atleast one row of data has the right age.

#### -> Hence by grouping Customer_ID, we are considering the minimum age as the correct age for the customer. By this way we can avoid large age values which may be Typo errors.

#### -> Values with special characters are converted to normal Integer values.

In [7909]:
df.Age

0          23
1          23
2        -500
3          23
4          23
         ... 
99995      25
99996      25
99997      25
99998      25
99999      25
Name: Age, Length: 100000, dtype: object

In [7910]:
df.Age.dtype

dtype('O')

In [7911]:
df.Age=df.Age.str.replace('-','').str.replace('_','')


In [7912]:
df.Age=pd.to_numeric(df.Age)

In [7913]:
df.Age.isna().sum()

0

In [7914]:
df.Age.dtype

dtype('int64')

In [7915]:
sample=df.merge(df.groupby('Customer_ID')['Age'].min().reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7916]:
df.Age=sample.Age_y*1

In [7917]:
df.Age.nunique()

43

In [7918]:
df.Age.sort_values(ascending=False)

51549    56
51548    56
51550    56
51551    56
51546    56
         ..
58274    14
58273    14
58272    14
4624     14
43788    14
Name: Age, Length: 100000, dtype: int64

# Occupation

#### -> Blank Value is Occupation is replaced as 'Unknown'

In [7919]:
df.Occupation.unique()

array(['Scientist', '_______', 'Teacher', 'Engineer', 'Entrepreneur',
       'Developer', 'Lawyer', 'Media_Manager', 'Doctor', 'Journalist',
       'Manager', 'Accountant', 'Musician', 'Mechanic', 'Writer',
       'Architect'], dtype=object)

In [7920]:
df.Occupation.nunique()

16

In [7921]:
df.Occupation=np.where(df.Occupation=='_______',np.NaN,df.Occupation)

In [7922]:
sample=df.merge(df.groupby('Customer_ID')['Occupation'].agg(pd.Series.mode).reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7923]:
df.Occupation.fillna(sample.Occupation_y,inplace=True)

In [7924]:
df.Occupation.unique()

array(['Scientist', 'Teacher', 'Engineer', 'Entrepreneur', 'Developer',
       'Lawyer', 'Media_Manager', 'Doctor', 'Journalist', 'Manager',
       'Accountant', 'Musician', 'Mechanic', 'Writer', 'Architect'],
      dtype=object)

# Annual Income

#### -> Annual_Income is converted to Numeric Datatype.
#### -> Few Customers have more than 1 Annual Income. So i have taken the median of the different Annual Incomes for each Customer ID to avoid the impact of Outliers.

In [7925]:
df.Annual_Income.dtype

dtype('O')

In [7926]:
df.Annual_Income.isna().sum()

0

In [7927]:
df.Annual_Income.value_counts(ascending=True)

14834.98_     1
31280.63_     1
12975.96_     1
45780.86_     1
9295469       1
             ..
33029.66     15
17816.75     15
36585.12     16
20867.67     16
17273.83     16
Name: Annual_Income, Length: 18940, dtype: int64

In [7928]:
df.Annual_Income=df.Annual_Income.str.replace('_','')

In [7929]:
df.Annual_Income=pd.to_numeric(df.Annual_Income)
df.Annual_Income.dtype

dtype('float64')

In [7930]:
df.groupby('Customer_ID')['Annual_Income'].median().reset_index()

Customer_ID  Annual_Income
0      CUS_0x1000      30625.940
1      CUS_0x1009      52312.680
2      CUS_0x100b     113781.390
3      CUS_0x1011      58918.470
4      CUS_0x1013      98620.980
...           ...            ...
12495   CUS_0xff3      17032.785
12496   CUS_0xff4      25546.260
12497   CUS_0xff6     117639.920
12498   CUS_0xffc      60877.170
12499   CUS_0xffd      41398.440

[12500 rows x 2 columns]

In [7931]:
sample=df.merge(df.groupby('Customer_ID')['Annual_Income'].median().reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7932]:
df.Annual_Income=sample.Annual_Income_y

In [7933]:
df.Annual_Income=np.round(df.Annual_Income)

In [7934]:
df.Annual_Income=df.Annual_Income.astype('int')

In [7935]:
df.Annual_Income.sort_values(ascending=False)

15688    179987
15695    179987
15689    179987
15690    179987
15691    179987
          ...  
39676      7006
39675      7006
80030      7006
39674      7006
80031      7006
Name: Annual_Income, Length: 100000, dtype: int64

In [7936]:
df.describe()

Age  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
count  100000.000000  100000.000000           84998.000000      100000.000000   
mean       33.024880   50505.125920            4194.170850          17.091280   
std        10.755849   38299.421504            3183.686167         117.404834   
min        14.000000    7006.000000             303.645417          -1.000000   
25%        24.000000   19342.750000            1625.568229           3.000000   
50%        33.000000   37000.000000            3093.745000           6.000000   
75%        41.000000   71683.250000            5957.448333           7.000000   
max        56.000000  179987.000000           15204.633330        1798.000000   

       Num_Credit_Card  Interest_Rate  Delay_from_due_date  \
count     100000.00000  100000.000000        100000.000000   
mean          22.47443      72.466040            21.068780   
std          129.05741     466.422621            14.860104   
min            0.00000       1.000000            -5.000000   
25%            4.00000       8.000000            10.000000   
50%            5.00000      13.000000            18.000000   
75%            7.00000      20.000000            28.000000   
max         1499.00000    5797.000000            67.000000   

       Num_Credit_Inquiries  Credit_Utilization_Ratio  Total_EMI_per_month  
count          98035.000000             100000.000000        100000.000000  
mean              27.754251                 32.285173          1403.118217  
std              193.177339                  5.116875          8306.041270  
min                0.000000                 20.000000             0.000000  
25%                3.000000                 28.052567            30.306660  
50%                6.000000                 32.305784            69.249473  
75%                9.000000                 36.496663           161.224249  
max             2597.000000                 50.000000         82331.000000

In [7937]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     84998 non-null   float64
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

# Monthly Inhand Salary

In [7938]:
df.Monthly_Inhand_Salary

0        1824.843333
1                NaN
2                NaN
3                NaN
4        1824.843333
            ...     
99995    3359.415833
99996    3359.415833
99997    3359.415833
99998    3359.415833
99999    3359.415833
Name: Monthly_Inhand_Salary, Length: 100000, dtype: float64

In [7939]:
df.Monthly_Inhand_Salary.isna().sum()

15002

In [7940]:
sample=df.merge(df.groupby('Customer_ID')['Monthly_Inhand_Salary'].median().reset_index(),left_on='Customer_ID',right_on='Customer_ID')


In [7941]:
df.Monthly_Inhand_Salary=sample.Monthly_Inhand_Salary_y*1

In [7942]:
df.Monthly_Inhand_Salary=np.round(df.Monthly_Inhand_Salary)

In [7943]:
df.Monthly_Inhand_Salary=df.Monthly_Inhand_Salary.astype('int')

In [7944]:
df.Monthly_Inhand_Salary.sort_values(ascending=False)

62735    15205
62734    15205
62728    15205
62729    15205
62730    15205
         ...  
83644      304
83645      304
83646      304
83647      304
83642      304
Name: Monthly_Inhand_Salary, Length: 100000, dtype: int64

In [7945]:
df.Monthly_Inhand_Salary.isna().sum()

0

# Num_Bank_Accounts , Num_Credit_Card, Num_Credit_Inquiries, Interest_Rate

#### -> Num_Bank_Accounts : 75th Percentile is 7 but  max value is 1798. 
#### -> Num_Credit_Card : 75th Percentile is 7 but  max value is 1499.
#### -> Interest_Rate : 75th Percentile is 20 but max value is 5797.
#### -> Num_Credit_Inquiries : 75th Percentile is 9 but max value is 2597.
#### -> Num_of Loan : 75th Percentile is 6 but max value is 1496.

### -> Outliers has to be taken care based on the requirement.

In [7946]:
df[['Num_Bank_Accounts','Num_Credit_Card','Interest_Rate','Num_Credit_Inquiries','Num_of_Loan']].describe()

Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_Credit_Inquiries
count      100000.000000     100000.00000  100000.000000          98035.000000
mean           17.091280         22.47443      72.466040             27.754251
std           117.404834        129.05741     466.422621            193.177339
min            -1.000000          0.00000       1.000000              0.000000
25%             3.000000          4.00000       8.000000              3.000000
50%             6.000000          5.00000      13.000000              6.000000
75%             7.000000          7.00000      20.000000              9.000000
max          1798.000000       1499.00000    5797.000000           2597.000000

## User Defined Functions to Calculate Upper Limit and Lower Limit

In [7947]:
def upper_limit(series):
    
    q1=series.quantile(0.25)
    q3=series.quantile(0.75)
    
    return q3+(1.5*(q3-q1))
    

In [7948]:
def lower_limit(series):
    
    q1=series.quantile(0.25)
    q3=series.quantile(0.75)
    
    if q1-(1.5*(q3-q1)) < 0 : 
        return 0
    else:
        
        return q1-(1.5*(q3-q1))
    

In [7949]:
def replace_values(x,upper_limit,lower_limit):
    

    
    if x>upper_limit:
       
        return upper_limit
    elif x<lower_limit:
       
        return lower_limit
    else:
    
        return x
        
   

## Num_Bank_Accounts, Num_Credit_Card, Num_Credit_Card

In [7950]:
df['Num_Bank_Accounts']= df['Num_Bank_Accounts'].apply(replace_values,args=[upper_limit(df.Num_Bank_Accounts),lower_limit(df.Num_Bank_Accounts)])

In [7951]:
df['Num_Credit_Card']= df['Num_Credit_Card'].apply(replace_values,args=[upper_limit(df.Num_Credit_Card),lower_limit(df.Num_Credit_Card)])

In [7952]:
df['Num_Credit_Inquiries']= df['Num_Credit_Inquiries'].apply(replace_values,args=[upper_limit(df.Num_Credit_Inquiries ),lower_limit(df.Num_Credit_Inquiries )])

In [7953]:
sample=df.merge(df.groupby('Customer_ID')[['Num_Bank_Accounts','Num_Credit_Card','Num_Credit_Inquiries']].max().reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7954]:
df.Num_Bank_Accounts=sample.Num_Bank_Accounts_y*1
df.Num_Credit_Card=sample.Num_Credit_Card_y*1
df.Num_Credit_Inquiries =sample.Num_Credit_Inquiries_y*1

In [7955]:
df.Num_Bank_Accounts=df.Num_Bank_Accounts.astype('int')
df.Num_Credit_Card=df.Num_Credit_Card.astype('int')
df.Num_Credit_Inquiries =df.Num_Credit_Inquiries.astype('int')

In [7956]:
df.Num_Bank_Accounts.unique()

array([ 3,  2,  1,  7,  4,  0,  8,  5,  6,  9, 10, 13, 11])

In [7957]:
df.Num_Credit_Card.unique()

array([ 4, 11,  5,  1,  7,  6,  8,  3,  9,  2, 10])

In [7958]:
df.Num_Credit_Inquiries.unique()

array([ 4,  2,  3,  9,  8,  5,  7,  1, 10, 18, 11, 13,  0, 12, 14,  6, 16,
       15, 17])

In [7959]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  int64  
 12  Num_of_Loan               100000 non-null  object 
 13  Type_of_Loan              88592 non-null   ob

## Interest Rate

In [7960]:
df.Interest_Rate.sort_values(ascending=False)

71641    5797
73583    5789
23512    5788
88505    5776
77710    5775
         ... 
47699       1
47700       1
47701       1
47702       1
76431       1
Name: Interest_Rate, Length: 100000, dtype: int64

In [7961]:
df['Interest_Rate']= df['Interest_Rate'].apply(replace_values,args=[upper_limit(df.Interest_Rate ),lower_limit(df.Interest_Rate)])

In [7962]:
df.Interest_Rate.sort_values(ascending=False)

7881     38.0
59488    38.0
20846    38.0
98349    38.0
20864    38.0
         ... 
47699     1.0
47700     1.0
47701     1.0
47702     1.0
76431     1.0
Name: Interest_Rate, Length: 100000, dtype: float64

In [7963]:
df['Interest_Rate']= df['Interest_Rate'].apply(replace_values,args=[upper_limit(df.Interest_Rate ),lower_limit(df.Interest_Rate)])

In [7964]:
df.Interest_Rate.sort_values(ascending=False)

7881     38.0
59488    38.0
20846    38.0
98349    38.0
20864    38.0
         ... 
47699     1.0
47700     1.0
47701     1.0
47702     1.0
76431     1.0
Name: Interest_Rate, Length: 100000, dtype: float64

In [7965]:
df.Interest_Rate.unique()

array([ 3.,  6.,  8.,  4.,  5., 38., 15.,  7., 12., 20.,  1., 14., 32.,
       16., 17., 10., 31., 25., 18., 19.,  9., 24., 13., 33., 11., 21.,
       29., 28., 30., 23., 34.,  2., 27., 26., 22.])

In [7966]:
df.Interest_Rate=df.Interest_Rate.astype('float')

# Num_of_Loan

In [7967]:
df.Num_of_Loan.sort_values(ascending=False)

80022      9_
391        9_
29172      9_
86707      9_
9211       9_
         ... 
41951    -100
41947    -100
77393    -100
94382    -100
77045    -100
Name: Num_of_Loan, Length: 100000, dtype: object

In [7968]:
df.Num_of_Loan=df.Num_of_Loan.str.replace('-','').str.replace('_','')

In [7969]:
df.Num_of_Loan=pd.to_numeric(df.Num_of_Loan)

In [7970]:
df[['Num_of_Loan']].describe()

Num_of_Loan
count  100000.000000
mean       10.761960
std        61.789933
min         0.000000
25%         2.000000
50%         3.000000
75%         6.000000
max      1496.000000

In [7971]:
df['Num_of_Loan']= df['Num_of_Loan'].apply(replace_values,args=[upper_limit(df.Num_of_Loan ),lower_limit(df.Num_of_Loan)])

In [7972]:
df.Num_of_Loan=df.Num_of_Loan.astype('int')

In [7973]:
df[['Num_of_Loan']].describe()

Num_of_Loan
count  100000.000000
mean        3.901870
std         2.951262
min         0.000000
25%         2.000000
50%         3.000000
75%         6.000000
max        12.000000

In [7974]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

## Payment_of_Min_Amount 

In [7975]:
df.Payment_of_Min_Amount.nunique()

3

In [7976]:
df.Payment_of_Min_Amount.unique()

array(['No', 'NM', 'Yes'], dtype=object)

In [7977]:
df.Payment_of_Min_Amount.value_counts()

Yes    52326
No     35667
NM     12007
Name: Payment_of_Min_Amount, dtype: int64

In [7978]:
df.Payment_of_Min_Amount=np.where(df.Payment_of_Min_Amount=='NM',np.NaN,df.Payment_of_Min_Amount)

In [7979]:
df.Payment_of_Min_Amount.unique()

array(['No', nan, 'Yes'], dtype=object)

In [7980]:
sample=df.merge(df.groupby('Customer_ID')['Payment_of_Min_Amount'].agg(pd.Series.mode).reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7981]:
df.Payment_of_Min_Amount.fillna(sample.Payment_of_Min_Amount_y,inplace=True)

In [7982]:
df.Payment_of_Min_Amount.unique()

array(['No', 'Yes'], dtype=object)

In [7983]:
df.Payment_of_Min_Amount.nunique()

2

## Credit_History_Age 

In [7984]:
df.Credit_History_Age.dtype

dtype('O')

In [7985]:
df.Credit_History_Age.unique()

array(['22 Years and 1 Months', nan, '22 Years and 3 Months',
       '22 Years and 4 Months', '22 Years and 5 Months',
       '22 Years and 6 Months', '22 Years and 7 Months',
       '26 Years and 7 Months', '26 Years and 8 Months',
       '26 Years and 9 Months', '26 Years and 10 Months',
       '26 Years and 11 Months', '27 Years and 0 Months',
       '27 Years and 1 Months', '27 Years and 2 Months',
       '17 Years and 9 Months', '17 Years and 10 Months',
       '17 Years and 11 Months', '18 Years and 1 Months',
       '18 Years and 2 Months', '18 Years and 3 Months',
       '18 Years and 4 Months', '17 Years and 3 Months',
       '17 Years and 4 Months', '17 Years and 5 Months',
       '17 Years and 6 Months', '17 Years and 7 Months',
       '17 Years and 8 Months', '30 Years and 8 Months',
       '30 Years and 9 Months', '30 Years and 10 Months',
       '30 Years and 11 Months', '31 Years and 0 Months',
       '31 Years and 1 Months', '31 Years and 2 Months',
       '31 Years and

In [7986]:
df.Credit_History_Age.fillna(0,inplace=True)

In [7987]:
def split_credit_history(value):
    
  
    
    if value!=0:
    
        li=value.split(' ')

        
        li[0]=int(li[0])
        li[3]=int(li[3])

        return np.round(((li[0]*12)+li[3])/12)
    else :
        return 0

In [7988]:
df.Credit_History_Age=df.Credit_History_Age.apply(split_credit_history)

In [7989]:
df.Credit_History_Age.nunique()

35

In [7990]:
sample=df.merge(df.groupby('Customer_ID')['Credit_History_Age'].mean().reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [7991]:
df.Credit_History_Age=np.round(sample.Credit_History_Age_y*1).astype('int')

In [7992]:
df.Credit_History_Age

0        17
1        17
2        17
3        17
4        17
         ..
99995    32
99996    32
99997    32
99998    32
99999    32
Name: Credit_History_Age, Length: 100000, dtype: int64

In [7993]:
df[['Credit_History_Age']].describe()

Credit_History_Age
count       100000.000000
mean            16.770720
std              7.826726
min              0.000000
25%             11.000000
50%             17.000000
75%             23.000000
max             33.000000

In [7994]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

## Delay from Due Date, Num_of_Delayed_Payment 

#### -> Delay from Due Date is clean.
#### -> Num of Delayed Payment is converted to numeric after removing typos and then the outliers are modified.
#### -> Missing Values in Num of Delayed Payment is replaced by median values

In [7995]:
df[['Delay_from_due_date']].describe()

Delay_from_due_date
count        100000.000000
mean             21.068780
std              14.860104
min              -5.000000
25%              10.000000
50%              18.000000
75%              28.000000
max              67.000000

In [7996]:
df[['Delay_from_due_date']]

Delay_from_due_date
0                        3
1                       -1
2                        3
3                        5
4                        6
...                    ...
99995                   23
99996                   18
99997                   27
99998                   20
99999                   18

[100000 rows x 1 columns]

In [7997]:
df.Num_of_Delayed_Payment.dtype


dtype('O')

In [7998]:
df.Num_of_Delayed_Payment=df.Num_of_Delayed_Payment.str.replace('_','')

In [7999]:
df.Num_of_Delayed_Payment=pd.to_numeric(df.Num_of_Delayed_Payment)

In [8000]:
df.Num_of_Delayed_Payment.sort_values()

61493   -3.0
23870   -3.0
71306   -3.0
59926   -3.0
86427   -3.0
        ... 
99973    NaN
99974    NaN
99992    NaN
99993    NaN
99998    NaN
Name: Num_of_Delayed_Payment, Length: 100000, dtype: float64

In [8001]:
df.Num_of_Delayed_Payment=np.where(df.Num_of_Delayed_Payment<0,0,df.Num_of_Delayed_Payment)

In [8002]:
df.Num_of_Delayed_Payment.sort_values()

35532    0.0
21593    0.0
21594    0.0
21595    0.0
8105     0.0
        ... 
99973    NaN
99974    NaN
99992    NaN
99993    NaN
99998    NaN
Name: Num_of_Delayed_Payment, Length: 100000, dtype: float64

In [8003]:
df[['Num_of_Delayed_Payment']].describe()

Num_of_Delayed_Payment
count            92998.000000
mean                30.934805
std                226.030273
min                  0.000000
25%                  9.000000
50%                 14.000000
75%                 18.000000
max               4397.000000

In [8004]:
df['Num_of_Delayed_Payment']= df['Num_of_Delayed_Payment'].apply(replace_values,args=[upper_limit(df.Num_of_Delayed_Payment ),lower_limit(df.Num_of_Delayed_Payment )])

In [8005]:
df[['Num_of_Delayed_Payment']].describe()

Num_of_Delayed_Payment
count            92998.000000
mean                13.469795
std                  6.464612
min                  0.000000
25%                  9.000000
50%                 14.000000
75%                 18.000000
max                 31.500000

In [8006]:
df['Num_of_Delayed_Payment'].isna().sum()

7002

In [8007]:
df['Num_of_Delayed_Payment'].fillna(df['Num_of_Delayed_Payment'].median(),inplace=True)

In [8008]:
df['Num_of_Delayed_Payment'].isna().sum()

0

In [8009]:
df['Num_of_Delayed_Payment']=df['Num_of_Delayed_Payment'].astype('int')

In [8010]:
df[['Num_of_Delayed_Payment']].describe()

Num_of_Delayed_Payment
count           100000.000000
mean                13.503240
std                  6.225163
min                  0.000000
25%                  9.000000
50%                 14.000000
75%                 18.000000
max                 31.000000

In [8011]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

# Changed_Credit_Limit

In [8012]:
df.Changed_Credit_Limit.sort_values(ascending=False)

66821        _
17400        _
2336         _
42935        _
84480        _
         ...  
62220    -0.02
10028    -0.01
56166    -0.01
35824    -0.01
56441    -0.01
Name: Changed_Credit_Limit, Length: 100000, dtype: object

In [8013]:
df.Changed_Credit_Limit.unique()

array(['11.27', '_', '6.27', ..., '27.38', '25.16', '21.17'], dtype=object)

In [8014]:
df[df.Changed_Credit_Limit=='0']

ID Customer_ID  Month         Name  Age          SSN  \
26707   0xb27d  CUS_0x8eaa  April  wayne coled   33  274-64-8413   
41779  0x10acd  CUS_0x1383  April      niklasb   18  000-58-0568   
56685  0x16223  CUS_0x2dc6   June  rian rohani   45  262-10-1868   
94085  0x23d47  CUS_0xbb33   June  christiaane   48  958-68-0188   

          Occupation  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  \
26707      Developer          16519                   1280                 10   
41779      Developer          80207                   6767                  5   
56685         Writer          80574                   6708                  0   
94085  Media_Manager         100972                   8286                  2   

       Num_Credit_Card  Interest_Rate  Num_of_Loan  \
26707                8           28.0            9   
41779                4           20.0            5   
56685                3            4.0            1   
94085                6            7.0            1   

                                            Type_of_Loan  Delay_from_due_date  \
26707  Not Specified, Not Specified, Auto Loan, Stude...                   16   
41779  Student Loan, Payday Loan, Not Specified, Payd...                   11   
56685                                        Payday Loan                   10   
94085                                       Student Loan                   12   

       Num_of_Delayed_Payment Changed_Credit_Limit  Num_Credit_Inquiries  \
26707                      18                    0                    18   
41779                      19                    0                    18   
56685                       2                    0                     2   
94085                       4                    0                     3   

      Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  \
26707          _          2504.31                 36.180186   
41779   Standard          1129.87                 41.517669   
56685       Good          1142.39                 24.946939   
94085          _          1138.38                 39.502356   

       Credit_History_Age Payment_of_Min_Amount  Total_EMI_per_month  \
26707                   9                   Yes            65.705179   
41779                   8                   Yes           260.418980   
56685                  30                    No            56.995653   
94085                  15                    No            48.697558   

      Amount_invested_monthly                 Payment_Behaviour  \
26707             26.73124272  High_spent_Medium_value_payments   
41779              256.718455   High_spent_Small_value_payments   
56685               __10000__   Low_spent_Medium_value_payments   
94085             1087.482887    Low_spent_Small_value_payments   

      Monthly_Balance  
26707     285.5208279  
41779     419.5512318  
56685     446.8304207  
94085             NaN

In [8015]:
df.Changed_Credit_Limit=np.where(df.Changed_Credit_Limit=='_',np.NaN,df.Changed_Credit_Limit)

In [8016]:
df.Changed_Credit_Limit=pd.to_numeric(df.Changed_Credit_Limit)

In [8017]:
df[['Changed_Credit_Limit']].describe()

Changed_Credit_Limit
count          97909.000000
mean              10.389025
std                6.789496
min               -6.490000
25%                5.320000
50%                9.400000
75%               14.870000
max               36.970000

In [8018]:
df.Changed_Credit_Limit.sort_values()

73391   -6.49
4241    -6.48
52876   -6.45
73476   -6.44
45171   -6.44
         ... 
99548     NaN
99618     NaN
99800     NaN
99892     NaN
99931     NaN
Name: Changed_Credit_Limit, Length: 100000, dtype: float64

In [8019]:
df.loc[99548]
df.loc[99618]
df.loc[99800]
df.loc[99892]
df.loc[99931]

ID                                                                    0x25f89
Customer_ID                                                        CUS_0xb11c
Month                                                                   April
Name                                                           yinka adegokej
Age                                                                        38
SSN                                                               546-94-4789
Occupation                                                            Manager
Annual_Income                                                           15320
Monthly_Inhand_Salary                                                    1461
Num_Bank_Accounts                                                           6
Num_Credit_Card                                                             7
Interest_Rate                                                            15.0
Num_of_Loan                                                     

In [8020]:
sample=df.merge(df.groupby('Customer_ID')['Changed_Credit_Limit'].median().reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [8021]:
df.Changed_Credit_Limit.fillna(sample.Changed_Credit_Limit_y,inplace=True)

In [8022]:
df.Changed_Credit_Limit.value_counts()

 8.22     139
 11.50    128
 11.32    127
 10.06    124
 7.35     123
         ... 
-4.57       1
-4.82       1
 31.59      1
 27.79      1
 21.17      1
Name: Changed_Credit_Limit, Length: 3634, dtype: int64

# Credit_Mix

In [8023]:
df.Credit_Mix.unique()

array(['_', 'Good', 'Standard', 'Bad'], dtype=object)

In [8024]:
df.head(2)

ID Customer_ID     Month           Name  Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  aaron maashoh   23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  aaron maashoh   23  821-00-0265  Scientist   

   Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0          19114                   1825                  3                4   
1          19114                   1825                  3                4   

   Interest_Rate  Num_of_Loan  \
0            3.0            4   
1            3.0            4   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                   -1   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                       7                 11.27                     4   
1                      14                 11.27                     4   

  Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0          _           809.98                  26.82262                  17   
1       Good           809.98                  31.94496                  17   

  Payment_of_Min_Amount  Total_EMI_per_month Amount_invested_monthly  \
0                    No            49.574949             80.41529544   
1                    No            49.574949             118.2802216   

                 Payment_Behaviour Monthly_Balance  
0  High_spent_Small_value_payments     312.4940887  
1   Low_spent_Large_value_payments     284.6291625

In [8025]:
df.Credit_Mix=np.where(df.Credit_Mix=='_',np.NaN,df.Credit_Mix)

In [8026]:
sample=df.merge(df.groupby('Customer_ID')['Credit_Mix'].agg(pd.Series.mode).reset_index(),left_on='Customer_ID',right_on='Customer_ID')

In [8027]:
df.Credit_Mix.fillna(sample.Credit_Mix_y,inplace=True)

In [8028]:
df.head(2)

ID Customer_ID     Month           Name  Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  aaron maashoh   23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  aaron maashoh   23  821-00-0265  Scientist   

   Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0          19114                   1825                  3                4   
1          19114                   1825                  3                4   

   Interest_Rate  Num_of_Loan  \
0            3.0            4   
1            3.0            4   

                                        Type_of_Loan  Delay_from_due_date  \
0  Auto Loan, Credit-Builder Loan, Personal Loan,...                    3   
1  Auto Loan, Credit-Builder Loan, Personal Loan,...                   -1   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Num_Credit_Inquiries  \
0                       7                 11.27                     4   
1                      14                 11.27                     4   

  Credit_Mix Outstanding_Debt  Credit_Utilization_Ratio  Credit_History_Age  \
0       Good           809.98                  26.82262                  17   
1       Good           809.98                  31.94496                  17   

  Payment_of_Min_Amount  Total_EMI_per_month Amount_invested_monthly  \
0                    No            49.574949             80.41529544   
1                    No            49.574949             118.2802216   

                 Payment_Behaviour Monthly_Balance  
0  High_spent_Small_value_payments     312.4940887  
1   Low_spent_Large_value_payments     284.6291625

In [8029]:
df.Credit_Mix.unique()

array(['Good', 'Standard', 'Bad'], dtype=object)

## Payment_Behaviour 

In [8030]:
df.Payment_Behaviour.unique()

array(['High_spent_Small_value_payments',
       'Low_spent_Large_value_payments',
       'Low_spent_Medium_value_payments',
       'Low_spent_Small_value_payments',
       'High_spent_Medium_value_payments', '!@9#%8',
       'High_spent_Large_value_payments'], dtype=object)

In [8031]:
df.Payment_Behaviour=np.where(df.Payment_Behaviour=='!@9#%8','Others',df.Payment_Behaviour)

In [8032]:
df.Payment_Behaviour.unique()

array(['High_spent_Small_value_payments',
       'Low_spent_Large_value_payments',
       'Low_spent_Medium_value_payments',
       'Low_spent_Small_value_payments',
       'High_spent_Medium_value_payments', 'Others',
       'High_spent_Large_value_payments'], dtype=object)

## Outstanding_Debt 

In [8033]:
df.Outstanding_Debt.unique()

array(['809.98', '605.03', '1303.01', ..., '3571.7_', '3571.7', '502.38'],
      dtype=object)

In [8034]:
df.Outstanding_Debt=df.Outstanding_Debt.str.replace('_','')

In [8035]:
df.Outstanding_Debt=pd.to_numeric(df.Outstanding_Debt)

In [8036]:
df.Outstanding_Debt.unique()

array([ 809.98,  605.03, 1303.01, ...,  620.64, 3571.7 ,  502.38])

In [8037]:
df.Outstanding_Debt.sort_values()

96652       0.23
96648       0.23
96649       0.23
96650       0.23
96651       0.23
          ...   
8740     4998.07
8741     4998.07
8742     4998.07
8737     4998.07
8743     4998.07
Name: Outstanding_Debt, Length: 100000, dtype: float64

In [8038]:
df[['Outstanding_Debt']].describe()

Outstanding_Debt
count     100000.000000
mean        1426.220376
std         1155.129026
min            0.230000
25%          566.072500
50%         1166.155000
75%         1945.962500
max         4998.070000

## Credit_Utilization_Ratio 

In [8039]:
df.Credit_Utilization_Ratio.unique()

array([26.82261962, 31.94496006, 28.60935202, ..., 41.25552226,
       33.63820798, 34.19246265])

In [8040]:
df.Credit_Utilization_Ratio.sort_values()

15860    20.000000
54207    20.100770
3580     20.172942
14319    20.244130
63420    20.257073
           ...    
87595    49.064277
62954    49.254983
17029    49.522324
68000    49.564519
9382     50.000000
Name: Credit_Utilization_Ratio, Length: 100000, dtype: float64

In [8041]:
df[['Credit_Utilization_Ratio']].describe()

Credit_Utilization_Ratio
count             100000.000000
mean                  32.285173
std                    5.116875
min                   20.000000
25%                   28.052567
50%                   32.305784
75%                   36.496663
max                   50.000000

In [8042]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

## Total_EMI_per_month

In [8043]:
df.Total_EMI_per_month.sort_values()

69229        0.0
52825        0.0
52826        0.0
52827        0.0
52828        0.0
          ...   
51614    82193.0
3084     82204.0
29514    82236.0
15300    82256.0
87013    82331.0
Name: Total_EMI_per_month, Length: 100000, dtype: float64

In [8044]:
df[['Total_EMI_per_month']].describe()

Total_EMI_per_month
count        100000.000000
mean           1403.118217
std            8306.041270
min               0.000000
25%              30.306660
50%              69.249473
75%             161.224249
max           82331.000000

In [8045]:
df['Total_EMI_per_month']= df['Total_EMI_per_month'].apply(replace_values,args=[upper_limit(df.Total_EMI_per_month ),lower_limit(df.Total_EMI_per_month )])

In [8046]:
df[['Total_EMI_per_month']].describe()

Total_EMI_per_month
count        100000.000000
mean            108.054040
std             104.531770
min               0.000000
25%              30.306660
50%              69.249473
75%             161.224249
max             357.600632

In [8047]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

##  Amount_invested_monthly

In [8048]:
df.Amount_invested_monthly.sort_values()

38871      0
78904      0
38421      0
78908      0
14177      0
        ... 
99862    NaN
99872    NaN
99925    NaN
99961    NaN
99974    NaN
Name: Amount_invested_monthly, Length: 100000, dtype: object

In [8049]:
df.Amount_invested_monthly=df.Amount_invested_monthly.str.replace('_','')

In [8050]:
df.Amount_invested_monthly=pd.to_numeric(df.Amount_invested_monthly)

In [8051]:
df[['Amount_invested_monthly']].describe()

Amount_invested_monthly
count             95521.000000
mean                637.412998
std                2043.319327
min                   0.000000
25%                  74.534002
50%                 135.925681
75%                 265.731733
max               10000.000000

In [8052]:
df.Amount_invested_monthly.fillna(df.Amount_invested_monthly.median(),inplace=True)

In [8053]:
df.Amount_invested_monthly=np.round(df.Amount_invested_monthly).astype('int')

In [8054]:
df[['Amount_invested_monthly']].describe()

Amount_invested_monthly
count             100000.00000
mean                 614.95518
std                 1999.72578
min                    0.00000
25%                   77.00000
50%                  136.00000
75%                  255.00000
max                10000.00000

## Monthly Balance 

In [8055]:
df.Monthly_Balance.sort_values()

71453    0.007759665
43200    0.088627865
77405    0.095482496
60346    0.131135651
69129    0.366147079
            ...     
99820            NaN
99839            NaN
99852            NaN
99854            NaN
99927            NaN
Name: Monthly_Balance, Length: 100000, dtype: object

In [8056]:
df.Monthly_Balance=df.Monthly_Balance.str.replace('_','')

In [8057]:
df.Monthly_Balance=pd.to_numeric(df.Monthly_Balance,errors='coerce')

In [8058]:
df[['Monthly_Balance']].describe()

Monthly_Balance
count     9.880000e+04
mean     -3.036437e+22
std       3.181295e+24
min      -3.333333e+26
25%       2.700922e+02
50%       3.367192e+02
75%       4.702202e+02
max       1.602041e+03

In [8059]:
df['Monthly_Balance']= df['Monthly_Balance'].apply(replace_values,args=[upper_limit(df.Monthly_Balance ),lower_limit(df.Monthly_Balance )])

In [8060]:
df[['Monthly_Balance']].describe()

Monthly_Balance
count     98800.000000
mean        388.298393
std         175.190838
min           0.000000
25%         270.092209
50%         336.719190
75%         470.220186
max         770.412152

In [8061]:
df.Monthly_Balance.fillna(df.Monthly_Balance.median(),inplace=True)

In [8062]:
df[['Monthly_Balance']].describe()

Monthly_Balance
count    100000.000000
mean        387.679442
std         174.227054
min           0.000000
25%         270.893427
50%         336.719190
75%         467.670597
max         770.412152

## Type of Loan 

In [8063]:
df.Type_of_Loan

0        Auto Loan, Credit-Builder Loan, Personal Loan,...
1        Auto Loan, Credit-Builder Loan, Personal Loan,...
2        Auto Loan, Credit-Builder Loan, Personal Loan,...
3        Auto Loan, Credit-Builder Loan, Personal Loan,...
4        Auto Loan, Credit-Builder Loan, Personal Loan,...
                               ...                        
99995                          Auto Loan, and Student Loan
99996                          Auto Loan, and Student Loan
99997                          Auto Loan, and Student Loan
99998                          Auto Loan, and Student Loan
99999                          Auto Loan, and Student Loan
Name: Type_of_Loan, Length: 100000, dtype: object

In [8064]:
df.Type_of_Loan.isna().sum()

11408

In [8065]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 27 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   ID                        100000 non-null  object 
 1   Customer_ID               100000 non-null  object 
 2   Month                     100000 non-null  object 
 3   Name                      100000 non-null  object 
 4   Age                       100000 non-null  int64  
 5   SSN                       100000 non-null  object 
 6   Occupation                100000 non-null  object 
 7   Annual_Income             100000 non-null  int64  
 8   Monthly_Inhand_Salary     100000 non-null  int64  
 9   Num_Bank_Accounts         100000 non-null  int64  
 10  Num_Credit_Card           100000 non-null  int64  
 11  Interest_Rate             100000 non-null  float64
 12  Num_of_Loan               100000 non-null  int64  
 13  Type_of_Loan              88592 non-null   ob

In [8066]:
df.to_csv('../Data/Clean_Data.csv')